DSEGOG-174 saving plots to redux store

sam-glendenning opened the pull request:
## Description
This allows for the user to save the plot they have open in a window to the Redux store. This lasts while the main window is not closed or refreshed, since sessions have yet to be setup.

The Save As button has been removed as I didn't deem it useful. Instead, the Save button saves the plot under the title it currently has. If a plot is renamed and saved, a new saved plot is stored in Redux, with the old copy remaining under the old name.

When the user saves a plot and closes it, it can be reopened with its state restored from Redux.

## Testing instructions
Try saving a plot and reopening it with the Edit button. All state should be restored.

- [x] Review code
- [x] Check Actions build
- [x] Review changes to test coverage

## Agile board tracking
Closes DSEGOG-174

Commit:
DSEGOG-174 plot window component now receives initial state from open plots component

Open plots component now fetches the plot config from state and passes it to each plot window. This means that, once we have state in the plot slice for each plot, it will be fetched and the plot can be opened with its saved state

Commit:
DSEGOG-174 plot config saving to redux store

Some settings aren't saving yet, need to rejig the code to make that happen

Commit:
DSEGOG-174 setting minima and maxima from state

Commit:
DSEGOG-174 fixing read-only plot channel bug

By mistakenly shallow-copying the selected plot channel array, some inner elements contained references to the old copy which could not be changed. Fixed this by deep copying the array

Commit:
DSEGOG-174 using array deep copy where required

This makes tests easier and allows for more consistent code

Commit:
Merge branch 'develop' into DSEGOG-174-saving-plot-sessions

Commit:
DSEGOG-174 adding deep copy for y-axis selection

Commit:
DSEGOG-174 fixing deep copy method

Commit:
DSEGOG-174 plot colours are now set from plot config state

Previously, these were reset whenever the plot was opened. Now they can be restored properly

Commit:
DSEGOG-174 preventing plot from opening on save

If a plot was saved under a different name, it was being opened again in a new window because I set open to true. Changing this to prevent this behaviour from happening

sam-glendenning commented:
Reminder for Sam to try and add the plot window position on screen to the plot saved state

Commit:
DSEGOG-174 updating some unit tests and refactoring test data

Commit:
Merge branch 'develop' into DSEGOG-174-saving-plot-sessions

Commit:
DSEGOG-174 adding spanGaps: true

With the addition of null points into the graph, we can get breaks in line charts. By default, Chart.js does not join these points with a line, meaning if there are two points separated by a null point, the line will break between them.

To bring back a continuous line, I have re-specified spanGaps: true in the Chart.js config. Later, we will add functionality where this option can be amended by the user, in order to support overnight gaps being visualised

Commit:
DSEGOG-174 updating records and plotSlice tests

Commit:
DSEGOG-174 updating snapshots and tests with filters applied

Commit:
DSEGOG-174 reverting snapshot changes and removing null data

I thought it best to tackle what we'll do with null data when we do the overnight gaps issue. This will make sure everything is consistent. As such, for now, no null data points will be inserted into the chart, as is already the case

Commit:
DSEGOG-174 bumping coverage

sam-glendenning commented:
One thing I overlooked - I forgot to implement functionality to restore the window position as it was. Is this still a desired feature?

louise-davies commented:
@sam-glendenning This is a feature on the wishlist & is part of one of the sessions requirements so I think it's still desired yes

louise-davies commented on src/plotting/util.tsx:
```diff
@@ -0,0 +1,13 @@
+import { SelectedPlotChannel } from '../app.types';
+
+// TODO test this
+export const deepCopySelectedPlotChannels = (
```
If you want to actually deep copy - you're best using `JSON.stringify` and `JSON.parse` - this method works fine as long as the options in `options` are primitives and not objects themselves (as otherwise those underlying objects are still the same by reference)

sam-glendenning replied:
That was an alternative I considered. Happy to make the change

louise-davies replied:
It's probably best to make it to prevent any future accident problems by the fact that this is only a single level deep copy. I had similar problems with the 2 y axis e2e test where I was accidentally modifying the underlying mock request data instead of deep copying because I just did object spread instead of JSON stringify&parse

sam-glendenning replied:
Resolved in https://github.com/ral-facilities/operationsgateway/pull/102/commits/19fbd764a1ca526d887c958298bcb5adcc5454c7

louise-davies reviewed (commented):
All the plot functionality still seems to work and saving a plot works and reopening that plot renders the same plot. I'm just not sure of the UX of "creating" a new plot automatically when the user saves a plot - especially in the case when it changes from `Untitled X` to a named plot.

I wonder if we do bring back the `Save As` button which does what the current Save button does, and have the `Save` button overwrite the plot config that was used to open the plot... Or have a dialogue pop up asking you whether you want to overwrite or create a new plot...

However, I'm happy for it to go in as is and we can change this later if needed, as I'm aware that Matt Jones is on leave so we can't ask him until next week

sam-glendenning commented:
I wasn't sure of this either. Any approach can be done, I'm just completely unsure what the user might want. I went with this as it was the safe option in case the user didn't want to lose an old plot, plus it acts as a method to "copy" plots in a sense? Anyway, shall we tag Matt and ask his opinion for when he does come back?

Commit:
DSEGOG-174 improving deep copy method

JSON.parse with JSON.stringify is more reliable in this case

louise-davies reviewed (dismissed):
Looks good. Just some questions about the save method, but we can change that in a later PR if needed. @mattjSTFC ping for when you get back but we want to know how the save plot behaviour is expected to work

sam-glendenning commented:
You will likely have to re-approve, I'm in the middle of trying to save the window position as well. I'll let you know

Commit:
DSEGOG-174 saving window width, height and position

Works fairly well? Especially on a single monitor. With multiple monitors, we can run into some inconsistencies but nothing that causes the window position to break.

Still looking at fixing a bug where the window position is only saved every second time the user presses the save button

Commit:
DSEGOG-174 extracting window size and position on save only

Rather than trying to coordinate state between the two components, I've attached a ref to the plot window to extract its size and position through that rather than through the component's componentDidUpdate method. This was inconsistent as to when it fired so this method is an improvement

Commit:
DSEGOG-174 globalising default window vars, fixing tests

Commit:
DSEGOG-174 updating snapshot

sam-glendenning commented:
The window size and position are now also saved when the save button is clicked. It works _fairly_ well, certainly on a single monitor, although on multiple monitors your mileage may vary. If you have any ideas on how to improve this, fire away although I think it will be challenging.

louise-davies commented:
@sam-glendenning I did say when this requirement was discussed that we could only do our best and that MMV depending on browsers/monitor setups, so this looks fine to me

sam-glendenning commented:
@louise-davies are you happy to approve?

louise-davies reviewed (approved):
As you say - the window position saving works pretty well if the monitor you have the OG main window open on is the same as the one you have the plots on. Otherwise it uses the wrong screen (in Chrome) or uses the wrong aspect ratios (FF) - but I think this is acceptable because a) there's literally not much else we can do re: controlling the size and position of the popup window and b) it's better than nothing, especially in the single window case.

Again, there's some alternative ideas on how saving plots should work re: overwriting existing plots vs creating new plots, but we can address those later

sam-glendenning commented:
Thanks for the extended testing on other browsers, I admit I had not done that. Ultimately, we'll see what end users think!
